File: abc_training/E/E_Erasing_Vertices_2.py
#          E - Erasing Vertices 2         
# ----------------------------------------
# 問題
# https://atcoder.jp/contests/abc267/tasks/abc267_e
# ----------------------------------------

from collections import deque

N, M = map(int, input().split())
A = list(map(int, input().split()))
G = [[] for _ in range(N)]
COST = [0] * N
for _ in range(M):
    u, v = map(int, input().split())
    u -= 1; v -= 1
    G[u].append(v)
    G[v].append(u)
    COST[u] += A[v]
    COST[v] += A[u]

def can(m):
    cost = COST[:]
    
    is_delete = [False] * N
    que = deque()
    for u in range(N):
        if cost[u] <= m:
            que.append(u)

    while que:
        u = que.popleft()
        for v in G[u]:
            cost[v] -= A[u]

            if not is_delete[v] and cost[v] <= m:
                is_delete[v] = True
                que.append(v)
    
    return all(is_delete)

l, r = -1, 1001001001001
while (r - l) > 1:
    mid = (l + r) // 2
    if can(mid):
        r = mid
    else:
        l = mid

print(r)

File: abc_training/E/E_Erasing_Vertices_2.rs
//          E - Erasing Vertices 2         
// ----------------------------------------
// 問題
// https://atcoder.jp/contests/abc267/tasks/abc267_e
// ----------------------------------------

// attributes
#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(non_snake_case)]

// imports
use std::collections::{HashMap, BTreeMap, VecDeque};

// input macro
// [Rustで競技プログラミング スターターキット](https://qiita.com/hatoo@github/items/fa14ad36a1b568d14f3e)
macro_rules! get {
    ($t:ty) => {
        {
            let mut line = String::new();
            std::io::stdin().read_line(&mut line).unwrap();
            line.trim().parse::<$t>().unwrap()
        }
    };
    ($($t:ty),*) => {
        {
            let mut line = String::new();
            std::io::stdin().read_line(&mut line).unwrap();
            let mut iter = line.split_whitespace();
            (
                $(iter.next().unwrap().parse::<$t>().unwrap(),)*
            )
        }
    };
    ($t:ty ; $n:expr) => {
        (0..$n).map(|_|
            get!($t)
        ).collect::<Vec<_>>()
    };
    ($($t:ty),* ; $n:expr) => {
        (0..$n).map(|_|
            get!($($t),*)
        ).collect::<Vec<_>>()
    };
    ($t:ty ;;) => {
        {
            let mut line = String::new();
            std::io::stdin().read_line(&mut line).unwrap();
            line.split_whitespace()
                .map(|t| t.parse::<$t>().unwrap())
                .collect::<Vec<_>>()
        }
    };
    ($t:ty ;; $n:expr) => {
        (0..$n).map(|_|
            get!($t ;;)
        ).collect::<Vec<_>>()
    };
}


// solve
fn main() {
    let (N, M) = get!(usize, usize);
    let A = get!(usize;;);
    let mut G = vec![vec![]; N];
    for _ in 0..M {
        let (mut u, mut v) = get!(usize, usize);
        u -= 1; v -= 1;
        G[u].push(v);
        G[v].push(u);
    }

    let can = |m: usize| -> bool {
        let mut cost = vec![0; N];
        for u in 0..N {
            for &v in &G[u] {
                cost[u] += A[v];
            }
        }
        let mut is_delete = vec![false; N];
        let mut que = VecDeque::new();
        for i in 0..N {
            if cost[i] <= m {
                que.push_back(i);
            }
        }
        while !que.is_empty() {
            let u = que.pop_front().unwrap();
            is_delete[u] = true;
            for &v in &G[u] {
                if !is_delete[v] {
                    cost[v] = if cost[v] <= A[u] { 0 } else { cost[v] - A[u] };
                    if cost[v] <= m {
                        que.push_back(v);
                    }
                }
            }
        }

        is_delete
            .iter()
            .fold(true, |acc, x| acc & x)
    };

    let mut left = 0;
    let mut right = 1001001001001;
    while (right - left) > 1 {
        let mid = (left + right) / 2;
        if can(mid) {
            left = mid;
        } else {
            right = mid;
        }
    }

    println!("{}", right);
}

